File: esports-backend/src/main/java/com/esports/service/FreelancerService.java
package com.esports.service;

import com.esports.dto.ServiceResponse;
import com.esports.entity.Service;
import com.esports.entity.User;
import com.esports.repository.ServiceRepository;
import com.esports.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@org.springframework.stereotype.Service
public class FreelancerService {

    @Autowired
    private ServiceRepository serviceRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private S3Service s3Service;

    @Transactional
    public Service createService(String email, String title, String description, Double price,
                               Integer deliveryTime, String category, Boolean active,
                               MultipartFile photo1, MultipartFile photo2, MultipartFile photo3) {

        User freelancer = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));

        Service service = new Service();
        service.setTitle(title);
        service.setDescription(description);
        service.setPrice(price);
        service.setCategory(category);
        service.setDeliveryTime(deliveryTime);
        service.setFreelancer(freelancer);
        service.setActive(active);

        // Handle photo uploads
        uploadPhoto(photo1, service::setPhoto1Url);
        uploadPhoto(photo2, service::setPhoto2Url);
        uploadPhoto(photo3, service::setPhoto3Url);

        return serviceRepository.save(service);
    }

    public List<ServiceResponse> getFreelancerServices(String email) {
        User freelancer = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));

        List<Service> services = serviceRepository.findByFreelancer(freelancer);

        return services.stream()
            .map(service -> new ServiceResponse(
                service.getId(),
                service.getTitle(),
                service.getDescription(),
                service.getPrice(),
                service.getDeliveryTime(),
                service.getCategory(),
                service.getActive(),
                service.getFreelancer().getName(),
                service.getCreatedAt(),
                service.getUpdatedAt(),
                generatePhotoUrl(service.getPhoto1Url()),
                generatePhotoUrl(service.getPhoto2Url()),
                generatePhotoUrl(service.getPhoto3Url())
            ))
            .collect(Collectors.toList());
    }

    @Transactional
    public Service updateService(String email, Long id, String title, String description, Double price,
                               Integer deliveryTime, String category, Boolean active,
                               MultipartFile photo1, MultipartFile photo2, MultipartFile photo3) {

        User freelancer = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));

        Service service = serviceRepository.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Service not found"));

        if (!service.getFreelancer().getId().equals(freelancer.getId())) {
            throw new SecurityException("Unauthorized access to service");
        }

        service.setTitle(title);
        service.setDescription(description);
        service.setPrice(price);
        service.setCategory(category);
        service.setDeliveryTime(deliveryTime);
        service.setActive(active);

        // Handle photo uploads
        if (photo1 != null && !photo1.isEmpty()) {
            uploadPhoto(photo1, service::setPhoto1Url);
        }
        if (photo2 != null && !photo2.isEmpty()) {
            uploadPhoto(photo2, service::setPhoto2Url);
        }
        if (photo3 != null && !photo3.isEmpty()) {
            uploadPhoto(photo3, service::setPhoto3Url);
        }

        return serviceRepository.save(service);
    }

    public Service getService(Long id) {
        return serviceRepository.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Service not found"));
    }

    @Transactional
    public void deleteService(String email, Long id) {
        User freelancer = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));

        Service service = serviceRepository.findById(id)
            .orElseThrow(() -> new NoSuchElementException("Service not found"));

        if (!service.getFreelancer().getId().equals(freelancer.getId())) {
            throw new SecurityException("Unauthorized access to service");
        }

        serviceRepository.delete(service);
    }

    private void uploadPhoto(MultipartFile photo, PhotoUrlSetter setter) {
        if (photo != null && !photo.isEmpty()) {
            try {
                Map<String, String> result = s3Service.upload(photo, "service-photos");
                setter.setUrl(result.get("url"));
            } catch (Exception e) {
                // Log error silently, continue without photo
            }
        }
    }

    private String generatePhotoUrl(String storedUrl) {
        if (storedUrl == null) return null;
        try {
            return s3Service.generateDownloadUrl(extractS3Key(storedUrl));
        } catch (Exception e) {
            return null;
        }
    }

    private String extractS3Key(String url) {
        if (url != null && !url.startsWith("http")) {
            return url;
        }
        if (url != null && url.contains("service-photos/")) {
            String keyPart = url.substring(url.indexOf("service-photos/"));
            if (keyPart.contains("?")) {
                keyPart = keyPart.substring(0, keyPart.indexOf("?"));
            }
            return keyPart;
        }
        return url;
    }

    @FunctionalInterface
    private interface PhotoUrlSetter {
        void setUrl(String url);
    }
}

File: Upkart1/Services/FreelancerService.cs
using Microsoft.EntityFrameworkCore;
using Upkart1.Data;
using Upkart1.DTO;
using Upkart1.Entities;
using Upkart1.Entities.Enums;

namespace Upkart1.Services
{
    public class FreelancerService : IFreelancerService
    {
        private readonly ApplicationDbContext _context;
        private readonly IS3Service _s3Service;

        public FreelancerService(ApplicationDbContext context, IS3Service s3Service)
        {
            _context = context;
            _s3Service = s3Service;
        }

        public async Task<Service> CreateService(string email, string title, string description, double price,
                                   int? deliveryTime, string category, bool active,
                                   IFormFile? photo1, IFormFile? photo2, IFormFile? photo3)
        {
            var freelancer = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var service = new Service
            {
                Title = title,
                Description = description,
                Price = price,
                Category = category,
                DeliveryTime = deliveryTime,
                FreelancerId = freelancer.Id,
                Active = active
            };

            service.Photo1Url = await UploadPhoto(photo1);
            service.Photo2Url = await UploadPhoto(photo2);
            service.Photo3Url = await UploadPhoto(photo3);

            _context.Services.Add(service);
            _context.SaveChanges();
            return service;
        }

        public List<ServiceResponse> GetFreelancerServices(string email)
        {
            var freelancer = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var services = _context.Services
                .Where(s => s.FreelancerId == freelancer.Id)
                .ToList();

            return services.Select(service => new ServiceResponse(
                service.Id,
                service.Title,
                service.Description,
                service.Price,
                service.DeliveryTime ?? 0,
                service.Category,
                service.Active,
                service.Freelancer?.Name ?? freelancer.Name,
                service.CreatedAt,
                service.UpdatedAt,
                GeneratePhotoUrl(service.Photo1Url),
                GeneratePhotoUrl(service.Photo2Url),
                GeneratePhotoUrl(service.Photo3Url)
            )).ToList();
        }

        public async Task<Service> UpdateService(string email, long id, string title, string description, double price,
                                   int? deliveryTime, string category, bool active,
                                   IFormFile? photo1, IFormFile? photo2, IFormFile? photo3)
        {
            var freelancer = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var service = _context.Services.FirstOrDefault(s => s.Id == id)
                ?? throw new KeyNotFoundException("Service not found");

            if (service.FreelancerId != freelancer.Id)
                throw new UnauthorizedAccessException("Unauthorized access to service");

            service.Title = title;
            service.Description = description;
            service.Price = price;
            service.Category = category;
            service.DeliveryTime = deliveryTime;
            service.Active = active;

            if (photo1 != null && photo1.Length > 0)
                service.Photo1Url = await UploadPhoto(photo1);
            if (photo2 != null && photo2.Length > 0)
                service.Photo2Url = await UploadPhoto(photo2);
            if (photo3 != null && photo3.Length > 0)
                service.Photo3Url = await UploadPhoto(photo3);

            _context.SaveChanges();
            return service;
        }

        public Service GetService(long id)
        {
            return _context.Services.FirstOrDefault(s => s.Id == id)
                ?? throw new KeyNotFoundException("Service not found");
        }

        public void DeleteService(string email, long id)
        {
            var freelancer = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var service = _context.Services.FirstOrDefault(s => s.Id == id)
                ?? throw new KeyNotFoundException("Service not found");

            if (service.FreelancerId != freelancer.Id)
                throw new UnauthorizedAccessException("Unauthorized access to service");

            _context.Services.Remove(service);
            _context.SaveChanges();
        }

        private async Task<string?> UploadPhoto(IFormFile? photo)
        {
            if (photo != null && photo.Length > 0)
            {
                try
                {
                    var result = await _s3Service.Upload(photo, "service-photos");
                    return result["url"];
                }
                catch
                {
                    // Log error silently, continue without photo
                    return null;
                }
            }
            return null;
        }

        private string? GeneratePhotoUrl(string? storedUrl)
        {
            if (storedUrl == null) return null;
            try
            {
                return _s3Service.GenerateDownloadUrl(ExtractS3Key(storedUrl));
            }
            catch
            {
                return null;
            }
        }

        private string ExtractS3Key(string url)
        {
            if (url != null && !url.StartsWith("http"))
                return url;
            if (url != null && url.Contains("service-photos/"))
            {
                var keyPart = url.Substring(url.IndexOf("service-photos/"));
                if (keyPart.Contains("?"))
                    keyPart = keyPart.Substring(0, keyPart.IndexOf("?"));
                return keyPart;
            }
            return url;
        }
    }
}
